toxcore/network.c

```diff
@@ -1745,7 +1745,17 @@ char *net_new_strerror(int error)
 #else
     char *str = (char *)malloc(256);
 #ifdef _GNU_SOURCE
-    str = strerror_r(error, str, 256);
+    char *retstr = strerror_r(error, str, 256);
+
+    if (retstr != str) {
```

Comment:
Maybe we should do this differently: allocate a buffer on the stack and then copy it into `str`, unconditionally. That way we can avoid this non-portable pointer compare.

Reply:
I have no idea what you are saying. Why is this pointer comparison non-portable? How a stack buffer would help when `strerror_r()` is allowed to return an internal buffer instead of using our stack buffer, so we would have the same issue over again?

Reply:
Pointers can only be compared if they are pointing within the same allocation. E.g.:

```c
char a[2];
char b[2];
if (&a[0] == &a[1]) ... // OK
if (&a[0] == &b[0]) ... // UB

// Similarly:
char a[2];
char *b = malloc(2);
if (&a[0] == b) ... // UB

// Or equally:
char *a = malloc(2);
char *b = malloc(2);
if (a == b) ... // UB
```

We can avoid the problem by doing:

```c
#define STRERROR_BUF_SIZE 256
char buf[STRERROR_BUF_SIZE];
const char *retstr = strerror_r(error, buf, sizeof(buf));

const size_t len = strlen(retstr);
assert(len < STRERROR_BUF_SIZE);  // can this happen? probably need a test to make sure it can't
char *str = (char *)malloc(len);
if (str != nullptr) {
  memcpy(str, retstr, len + 1); // copy including null terminator
}
return str;
#undef STRERROR_BUF_SIZE
```

Reply:
If you want to avoid doing `retstr != str` comparison for some odd reason (how is comparing two `char *` is non-portable?), it can be removed as it is and the code would still be correct. There is no need to add any extra buffer. The comparison is in there just to avoid doing a useless `memmove(str, str, _)` in case `strerror_r()` stored the string in our buffer.